### PR TITLE
make some small adjustments to the i18n code

### PR DESCRIPTION
**Pull request by ivanhofer**
Hey Geoff,
I'm happy to see you did choose `typesafe-i18n` for the challenge :)

I glanced over the code and made some small adjustments.
Feel free to use some parts of it or just take it as an inspiration.

### CODE REVIEW

File: src/i18n/de/index.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
-import type { BaseTranslation } from '../i18n-types';
+import type { Translation } from '../i18n-types';
```
Comment (ivanhofer) [inline src/i18n/de/index.ts:1]: this brings typesafety to your translations ;)

File: src/i18n/de/index.ts [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
 
-const en: BaseTranslation = {
+const de: Translation = {
 	language: 'Deutsch',
 	happyHolidays: 'Schöne Ferien!',
-	christmasIsComing: '{ date|fullDay } ist in { time }',
-	day: '{{ ein Tag | ?? Tagen}}'
+	christmasIsComing: '{date|fullDay} ist in {time}',
```
Comment (ivanhofer) [inline src/i18n/de/index.ts:6]: arguments currently need to be trimmed to make `TypeScript` happy

File: src/i18n/en/index.ts [hunk 3]
```diff
@@ -4,7 +4,7 @@ const en: BaseTranslation = {
 	language: 'English',
 	happyHolidays: 'Happy Holidays!',
 	christmasIsComing: '{ date|fullDay } is in { time }',
-	day: '{{ one day | ?? days}}' // not sure if there's a way to name this parameter
+	day: '{{ days: one day | ?? days }}'
```
Comment (ivanhofer) [inline src/i18n/en/index.ts:7]: yes you can ;)

File: src/routes/day/8/+page.svelte [hunk 4]
```diff
@@ -16,9 +17,11 @@
 	const locales: Locales[] = ['en', 'de', 'ja-JP'];
 	let localeIdx = 0;
 
-	function changeLocale() {
+	async function changeLocale() {
 		localeIdx = (localeIdx + 1) % locales.length;
-		setLocale(locales[localeIdx]);
+		const newLocale = locales[localeIdx];
+		await loadLocaleAsync(newLocale);
```
Comment (ivanhofer) [inline src/routes/day/8/+page.svelte:23]: usually you would load the locales on demand

File: src/routes/day/8/+page.ts [hunk 5]
```diff
@@ -1,11 +1,12 @@
...
 
 import type { PageLoad } from './$types';
+import { baseLocale } from '$i18n/i18n-util.js';
 
 // based on https://github.com/ivanhofer/typesafe-i18n-demo-sveltekit example
 // not using layout loads because I only need it at this route
 export const load: PageLoad = async () => {
-	await loadAllLocalesAsync();
-	setLocale('en');
+	await loadLocaleAsync(baseLocale);
```
Comment (ivanhofer) [inline src/routes/day/8/+page.ts:10]: not sure if everything was intentially loaded together, or just an oversight

File: src/i18n/en/index.ts (hunk 3 above)
Comment (ivanhofer) [reply src/i18n/en/index.ts:7]: But I saw that there is a bug in the type generator.:
`day` in `RootTranslation` should be typed as `RequiredParams<'days'>` and not just `string`. It seems I didn't account for the case that just a `PluralPart` gets defined.
I'll open an issue in `typesafe-i18n`.

**Comment by geoffrich** [comment]
Ah, thanks! 

Will give this a closer look later, but while I have you here - one part of the challenge I couldn't quite figure out how to do in typesafe-i18n was [dynamic slots](https://github.com/Advent-Of-Vue/2022-i18n-happy-holidays/blob/4db5d15bc98ee82217a8e001803ffc1677b9fc04/src/App.vue#L16-L27) using vue-i18n's [i18n-t](https://vue-i18n.intlify.dev/guide/advanced/component.html#basic-usage) element. Would I just need to break up the translation strings manually or does typesafe-i18n have any similar feature? I'm assuming the former, since I don't know how I'd implement it as a component, but figured I'd ask 😄

**Comment by ivanhofer** [comment]
Yeah, since `typesafe-i18n` works with any framework, there are some things that it does not support out of the box. A pretty niche use-case in my opinion. It is also not quite easy to offer this for all different frameworks and staying fully typesafe.

Probably the best way to do this currently is like this:
```svelte
<p>
	{@html $LL.christmasIsComing({ 
			date: christmas,
			time: `<span style="color:green">${$LL.day({ days: daysUntilChristmas })}</span>`
		})
	}
</p>
```
You could also create a `wrapIntoGreenSpan` formatter, but then you would have UI information in your translations, which is a bad thing. 

If you want to spent some time, you could also convert [this example](https://github.com/ivanhofer/typesafe-i18n#how-do-i-render-a-component-inside-a-translation) to `Svelte`. I never actually tried since I never needed it. In terms of composition doing such abstract things is sometimes challengeing with `Svelte` :)

**Review by geoffrich — APPROVED** [review]
Thank you, this was very helpful!

**Comment by geoffrich** [comment]
Also, thanks for pointing me to that component example @ivanhofer - I was able to get something working using slot props https://github.com/geoffrich/advent-of-sveltekit-2022/commit/875a2ca11c387247a7ad40f2bb0c4839c5f791c4

**Comment by ivanhofer** [comment]
Looks good. I will shamelessly copy your `WrapTranslation` example and add it to the `typesafe-i18n` docs 😅.

**Pull request merged**